### PR TITLE
expressio: fix data race of rand function

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

```
WARNING: DATA RACE
Read at 0x00c000345508 by goroutine 81:
  math/rand.(*rngSource).Int63()
      /usr/lib/go/src/math/rand/rng.go:244 +0x8f
  github.com/pingcap/tidb/expression.(*builtinRandSig).evalReal()
      /usr/lib/go/src/math/rand/rand.go:85 +0x7a
  github.com/pingcap/tidb/expression.(*ScalarFunction).EvalReal()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/scalar_function.go:239 +0x73
  github.com/pingcap/tidb/expression.executeToReal()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/chunk_executor.go:143 +0x79
  github.com/pingcap/tidb/expression.evalOneColumn()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/chunk_executor.go:75 +0x2c0
  github.com/pingcap/tidb/expression.(*defaultEvaluator).run()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/evaluator.go:47 +0x664
  github.com/pingcap/tidb/expression.(*EvaluatorSuite).Run()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/evaluator.go:113 +0x113
  github.com/pingcap/tidb/executor.(*projectionWorker).run()
      /home/aliv/go/src/github.com/pingcap/tidb/executor/projection.go:356 +0x246

Previous write at 0x00c000345508 by goroutine 116:
  math/rand.(*rngSource).Int63()
      /usr/lib/go/src/math/rand/rng.go:244 +0xab
  github.com/pingcap/tidb/expression.(*builtinRandSig).evalReal()
      /usr/lib/go/src/math/rand/rand.go:85 +0x7a
  github.com/pingcap/tidb/expression.(*ScalarFunction).EvalReal()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/scalar_function.go:239 +0x73
  github.com/pingcap/tidb/expression.executeToReal()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/chunk_executor.go:143 +0x79
  github.com/pingcap/tidb/expression.evalOneColumn()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/chunk_executor.go:75 +0x2c0
  github.com/pingcap/tidb/expression.(*defaultEvaluator).run()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/evaluator.go:47 +0x664
  github.com/pingcap/tidb/expression.(*EvaluatorSuite).Run()
      /home/aliv/go/src/github.com/pingcap/tidb/expression/evaluator.go:113 +0x113
  github.com/pingcap/tidb/executor.(*projectionWorker).run()
      /home/aliv/go/src/github.com/pingcap/tidb/executor/projection.go:356 +0x246
```
### What is changed and how it works?

Add a mutex for `rand` to prevent data race.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/builtin_math.go [hunk 1]
```diff
@@ -988,19 +989,23 @@ func (c *randFunctionClass) getFunction(ctx sessionctx.Context, args []Expressio
 
 type builtinRandSig struct {
 	baseBuiltinFunc
+	mu      *sync.Mutex
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: All the projectWorkers will battle for this Mutex?

Comment (lamxTyler) [reply expression/builtin_math.go]: Yes, and this is the only way to make it correct.

**Pull request merged**